pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -0,0 +1,389 @@
+			deleting += 1
+		} else if !cd.Spec.Installed {
+			installing += 1
+		}
+	}
+	logger.WithFields(log.Fields{
+		"installing": installing,
+		"deleting":   deleting,
+		"total":      len(poolCDs),
+		"ready":      len(poolCDs) - installing - deleting,
+	}).Info("found clusters for ClusterPool")
+
+	// If too many, delete some.
+	// TODO: improve logic here, delete oldest, or delete still installing in favor of those that are ready
+	if len(poolCDs)-deleting > clp.Spec.Size {
+		deletionsNeeded := len(poolCDs) - deleting - clp.Spec.Size
+		if err := r.deleteExcessClusters(clp, poolCDs, deletionsNeeded, logger); err != nil {
+			return reconcile.Result{}, err
+		}
+	} else if len(poolCDs)-deleting < clp.Spec.Size {
+		// If too few, create new InstallConfig and ClusterDeployment.
+		if err := r.addClusters(clp, clp.Spec.Size-len(poolCDs)+deleting, logger); err != nil {
+			log.WithError(err).Error("error adding clusters")
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileClusterPool) addClusters(
+	clp *hivev1.ClusterPool,
+	newClusterCount int,
+	logger log.FieldLogger) error {
+	logger.Infof("Adding %d clusters", newClusterCount)
+
+	// Load the pull secret if available:
+
+	for i := 0; i < newClusterCount; i++ {
+		if err := r.createCluster(clp, logger); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcileClusterPool) createCluster(
+	clp *hivev1.ClusterPool,
+	logger log.FieldLogger) error {
+
+	ns, err := r.obtainRandomNamespace(clp)
+	if err != nil {
+		logger.WithError(err).Error("error obtaining random namespace")
+		return err
+	}
+	logger.Infof("Creating new cluster in namespace: %s", ns.Name)
+
+	// We will use this unique random namespace name for our cluster name.
+	builder := &clusterresource.Builder{
+		Name:        ns.Name,
+		Namespace:   ns.Name,
+		BaseDomain:  clp.Spec.BaseDomain,
+		DeleteAfter: clp.Spec.DeleteAfter.Duration.String(),
+		// TODO:
+		ReleaseImage:     "quay.io/openshift-release-dev/ocp-release:4.3.3-x86_64",
+		WorkerNodesCount: int64(3),
+		MachineNetwork:   "10.0.0.0/16",
+		ClusterPool:      types.NamespacedName{Namespace: clp.Namespace, Name: clp.Name},
+	}
+
+	// Load the pull secret if one is specified (may not be if using a global pull secret)
+	if clp.Spec.PullSecretRef != nil {
+		pullSec := &corev1.Secret{}
+		err := r.Client.Get(context.Background(),
+			types.NamespacedName{Namespace: clp.Namespace, Name: clp.Spec.PullSecretRef.Name},
+			pullSec)
+		if err != nil {
+			logger.WithError(err).Error("error reading pull secret")
+			return err
+		}
+		builder.PullSecret = string(pullSec.Data[".dockerconfigjson"])
+	}
+
+	// TODO: regions being ignored throughout, and not exposed in create-cluster cmd either
+	var credsSecretName string
+	if clp.Spec.Platform.AWS != nil {
+		credsSecretName = clp.Spec.Platform.AWS.CredentialsSecretRef.Name
+		// Lookup the platform creds secret for this pool which we are currently assuming
```

Comment:
Why are we assuming that these are in the hive namespace instead of the namespace of the ClusterPool?

Reply:
ClusterPool was originally non-Namespaced. I will address.